### PR TITLE
Add analytics sdk 0.1.0

### PR DESCRIPTION
**Pull request by colmsnowplow**
Some to-do's left as inline comments. I can move them to issues before release if we want to do that, or if we're happy to leave them in-line for early releases I'm happy with that also.

Design decisions documented in comments to make reviewing easier

A couple of things I'm still undecided upon:

- I used a pointer for timestamps, in order to allow nil as an output. Not sure if there's a better option.
- Separate functions for addition of geo_location - does this make things messier/less maintainable?
- What should the `Get` functions output where the field(s) exist but there's no value?

I'll still be tinkering with things in a separate branch, will merge it into this PR if I have improvements ready to add, but otherwise I think this is good enough for a beta/prototype, and we can release without including those experiments (with a view to making improvements in a subsequent release).

### CODE REVIEW

File: analytics/transform.go
Comment (paulboocock) [inline analytics/transform.go:232]: I'd usually pull "magic numbers" into a constant somewhere.

**Review by paulboocock — APPROVED** [review]
Looks like a decent 0.1.0 release to me. Made a few suggestions, take them or leave them but I'm happy to move forward with this. 
I imagine we will find a few bits and bobs as we start integrating it into other things but I think it looks solid for a first release.

File: analytics/transform.go
Comment (colmsnowplow) [reply analytics/transform.go:17]: Interesting - my test has it at about ~15% faster, but with easyjson at about ~25% faster.

However easyjson isn't as straightforward, and the unit tests would need to be fixed to accommodate. 

So I think the best thing to do is use jsoniter for now, but open an issue to explore easyjson and potentially other options when we have the opportunity to do so.

File: analytics/transform.go
Comment (colmsnowplow) [reply analytics/transform.go:186]: Sounds good. Just to call out the difference between the Get and GetSubset functions:

GetValue will return an error for GetValue, but no error for either GetSubset function... Should these also return errors if none of the subset of fields have a value?

File: analytics/transform.go
Comment (colmsnowplow) [reply analytics/transform.go:232]: So the other sdks would use len(enrichedEventFieldTypes). I don't think this is a terribly consequential decision so sticking it into a const and moving on. :)

File: analytics/transform.go
Comment (colmsnowplow) [reply analytics/transform.go:41]: Yup, good points both. I think I started with pointer for some reason that then became unimportant.

**Comment by paulboocock** [comment]
Flicking through the recent changes on my phone, what happens when a Timestamp is missing in the output (or any other non-required field that could be null in `tp2`)? 

As an example, let's say I send an event in which has no true_timestamp. 
Does the resulting json contain the key with the value of nil or is the entire key value pair missing in the output json?

What do the other analytics sdks do in this scenario? The two options are always have 131 json keys but some have null/nil values or only have the number of json keys which have values. I quite like the consistency of always have 131 keys so you can always guarantee you can read it but then that'll be a lot of nulls for many events.

**Comment by colmsnowplow** [comment]
I will confirm that this is actually what this sdk does (esp after changing the timestamp output), but the _desired_ behaviour is that the key doesn't exist at all. This is what the other sdks produce. At least, this is the behaviour when the input value is empty.

IMO it's better to output a smaller object than a large one with loads of fields, but I take the point on the consistency.

Perhaps including null fields should be a feature we consider?

**Comment by paulboocock** [comment]
That makes sense, I prefer the smaller output particularly when writing to other systems and thinking about storage and transfer. It's always easy to check for key existance. This was just a thought that popped into my head whilst reading through the changes as it differs from how you'd typically build a web API where you'd usually get a consistent response from a server but that's not what we're building here 😅

File: analytics/transform.go
Comment (paulboocock) [reply analytics/transform.go:174]: I think GetSubset is an interesting use case, for example if you ask for two values and you only find one, is that an error or an array with one element? If its the latter (which feels right) then an empty array seems more expected than an error.

Comment (colmsnowplow) [reply analytics/transform.go:174]: Yeah that's exactly what my thoughts are on it. :)

File: analytics/transform.go
Comment (paulboocock) [reply analytics/transform.go]: I guess we're on different CPUs/other hardware? Our laptops are a couple of generations apart so we're probably seeing some differences there too. I didn't test easyjson but interesting to know we might have another option to test that could potentially be even faster if we see bottlenecks.

Comment (colmsnowplow) [reply analytics/transform.go]: Perhaps yeah, tbh I've not been scientific about it - if we wanna take benchmarking results seriously I think doing it in CI is probably the best way to go.

For now it's all good, we can figure out what the genuine results are at a later point. :)

File: analytics/transform.go
Comment (colmsnowplow) [reply analytics/transform.go]: I've removed `parseNullableTime` and moved the code into `parseTime` - which still returns `nil, err` if it hits an error. I don't think this poses any issue, since the public methods which use it will all return `nil, err` if any of the parser functions hit an error, regardless of the parser's other return value.

File: analytics/transform.go
Comment (paulboocock) [inline analytics/transform.go:133]: It'd be nice to figure out how to extract this out of the functions, perhaps into a higher level object or something, so that you only need to split the event TSV once then you can call `Get*` and `To*` without needing to recompute the splitting.
I don't think this is required for 0.1.0, as for most use cases I think you'll call one of the methods in particular and then move on but for usecases where you want to call it twice then you won't want to be calling `Split` multiple times. Maybe worth filing an issue for 0.2.0, I think the only reason I'd be tempted to do it for 0.1.0 is I anticipate this will cause an API change.

**Review by paulboocock — APPROVED** [review]
This is looking pretty tidy now 👌

**Comment by paulboocock** [comment]
And one final comment, it'd be good to add a GitHub action which runs tests on pushes. You can probably steal most (all?) of the sql-runner one: https://github.com/snowplow/sql-runner/blob/master/.github/workflows/build.yml

File: analytics/transform.go
Comment (colmsnowplow) [reply analytics/transform.go:133]: I'm struggling a bit to conceptualise what you have in mind here. Perhaps we can chat about it whenever suits?

Comment (paulboocock) [reply analytics/transform.go:133]: I'm imagining something like:

```
var parsedEvent = parse(event) // This could be an interface/type to hide implementation details although ultimately is probably just the split string return.
parsedEvent.toJson();
parsedEvent.Get*();
```

Comment (colmsnowplow) [reply analytics/transform.go:133]: Ok I follow. I didn't realise you meant a change to the public interface rather than internals. 

I see what you mean now, will take a look. :)

**Comment by colmsnowplow** [comment]
OK GH action to test added.  No makefile or dockerisation, that all feels like overkill but if we need it I can add it.

Also we now have a ParseEvent() method and the Get*/To* functions are now methods on the ParsedEvent type, which ParseEvent returns.

My only niggling uncertainty is that I was hoping to abstract away all the length checks, but since technically ParseEvent doesn't have to be called, there's no codified guarantee of safety, so the Get* methods all do them manually, and mapifyGoodEvent (which is called by all the others) does so too.

**Comment by colmsnowplow** [comment]
Additionally: Feels like code-wise we're looking good, so I will find some time to create a README before release. Might want to hold off on a final pass/second reviewer until I've done that to save us some effort. :)

**Review by paulboocock — APPROVED** [review]
LGTM! Yep totally agree that this is looking pretty good now. Returning the slice is nice, gives options to go rogue and pick indexes directly off it (although we wouldn't recommend that!) or use the helper methods to be a bit safer :-) 
I think this will cover our usecases quite nicely. Might be worth tagging a `v0.1.0-beta.1` release of this to give it one last shake whilst we wait for the second reviewer.

**Comment by colmsnowplow** [comment]
Yeah for sure. TBH I'd be more than happy to leave it released as beta until we're close to ready to release the stream-replicator integration.

**Comment by colmsnowplow** [comment]
Closing in favour of https://github.com/snowplow/snowplow-golang-analytics-sdk/pull/3

**Pull request closed without merging**
